packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -928,25 +1004,88 @@ class TextPainter {
   /// called.
   void layout({ double minWidth = 0.0, double maxWidth = double.infinity }) {
-    assert(text != null, 'TextPainter.text must be set to a non-null value before using the TextPainter.');
-    assert(textDirection != null, 'TextPainter.textDirection must be set to a non-null value before using the TextPainter.');
-    // Return early if the current layout information is not outdated, even if
-    // _needsPaint is true (in which case _paragraph will be rebuilt in paint).
-    if (_paragraph != null && minWidth == _lastMinWidth && maxWidth == _lastMaxWidth) {
-      return;
+    assert(!maxWidth.isNaN);
+    assert(!minWidth.isNaN);
+    final InlineSpan? text = this.text;
+    if (text == null) {
+      throw StateError('TextPainter.text must be set to a non-null value before using the TextPainter.');
+    }
+    final TextDirection? textDirection = this.textDirection;
+    if (textDirection == null) {
+      throw StateError('TextPainter.textDirection must be set to a non-null value before using the TextPainter.');
+    }
+    final _TextPainterLayoutCache? cachedLayout = _layoutCache;
+    final double paintOffsetAlignment = _computePaintOffsetFraction(textAlign, textDirection);
+
+    // The assumption here is that if a Paragraph's width is already >= its
+    // maxIntrinsicWidth, further increasing the input width does not change its
+    // layout (but may change the paint offset if it's not left-aligned). This is
+    // true even for TextAlign.justify, since eot is also considered a hard
+    // line break so it will behave exactly the same as TextAlign.start when its
+    // width >= maxIntrinsicWidth.
+    //
+    // An exception to this is when the text is not left-aligned, and the input
+    // width is double.infinity. Since the resulting Paragraph will have a width
+    // of double.infinity, and to make the text visible the paintOffset.dx is
+    // bound to be double.negativeInfinity, which invalidates all arithmetic
+    // operations.
+    final bool needsLayout = cachedLayout == null
+      || (maxWidth != cachedLayout.layout.width
+        // Always needsLayout when the current paintOffset and the paragraph
+        // width are not finite: we won't be able to translate the paint offset
+        // using addition/subtraction.
+        && ((!cachedLayout.offset.dx.isFinite && !cachedLayout.paragraph.width.isFinite)
+          || cachedLayout.layout.width < cachedLayout.layout.maxIntrinsicLineExtent
+          || maxWidth < cachedLayout.layout.maxIntrinsicLineExtent));
+
+    final _TextLayout newLayout;
+    if (needsLayout) {
+      // Try to avoid laying out the paragraph with maxWith=double.infinity
+      // when the text is not left-aligned. When maxIntrinsicWidth is still
+      // unkown, use a relatively large number with decent precision instead of
+      // double.infinity.
+      const double largeWidth = 32768; // 2^15, mantissa precision: 2^15/2^52 ≅ 7.2759576e-12.
```

Comment:
This seems like it would have weird effects in a world where a significant transform has been applied?

Reply:
(e.g. scaling by 1e6 or something)

Reply:
maybe i misunderstood; it seems maybe this is just a performance optimization? if so we should make it clearer in the comments here that this number doesn't _really_ matter all that much, it's just a threshold after which we'll be doing more computation.

might be worth documenting the threshold as well, since if i understand correctly it'll double the cost of layout for already-expensive layouts, so the performance curve will have a sharp cliff for large texts?

Reply:
Removed this part of the code. I was told line breaking isn't that expensive so it's ok to call `layout` twice once in a while.